### ISSUE TITLE
test location is node_modules/gherkin-wrapper/.../playwright.js

### ISSUE THREAD

**Issue by cassus** [Niitch/gherkin-wrapper#4]
### Observed behavior
- In the playwright UI mode the tests show up under node_modules/gherkin-wrapper/lib/wrappers/playwright.js and not in the file where they were defined.
- The tests doesn't show up in the VSCode plugin at all.
- If I run `playwright:test -- fileName.ts` where fileName.ts has the `wrapper.test` no tests are run.

### Expected behavior
- [x] Option A (ideal): Playwright understood the location of the test as the feature file and line where the Example / Scenario was specified
  - [x] Works for the playwright test runner (visible in reports an UI mode) (_see [microsoft/playwright/23157](https://github.com/microsoft/playwright/issues/23157)_)
  - [ ] Works for the vscode test runner
- [x] Option B (workaround): Playwright understood the location of the test as the TS file where I called `wrapper.test`.

![Playwright_Test](https://github.com/Niitch/gherkin-wrapper/assets/316826/142d2c48-100d-42c5-a616-ab7a86310c18)

### Notes
Maybe this can be done if calling run the test.describe in the file where now we run `wrapper.test(`${__dirname}/XY.feature`)`

**Comment by cassus** [comment]
Workaround:

`const wrapper = createBddWrapper(test, (...args) => test(...args))`

With createBddWrapper as 
```ts

type TestFn = (
  ...args: Parameters<typeof baseTest>
) => ReturnType<typeof baseTest>

export function createBddWrapper<T extends typeof baseTest>(
  test: T,
  callTest: TestFn,
) {
  const testWrapper = Object.assign(callTest.bind({}), test) as T

  // eslint-disable-next-line new-cap -- external lib
  return new GherkinWrapper.forPlaywright(testWrapper)
}
```

**Comment by cassus** [comment]
Upgraded workaround also locates the describe blocks in the proper file

### Usage
```ts
createBddWrapper(test, {
    describe: (...args) => test.describe(...args),
    test: (...args) => test(...args),
  })
```

### Definition

```ts
import GherkinWrapper from "gherkin-wrapper"

import { test as baseTest } from "src/lib/playwright-test"

/**
 * Given callable type that has other properties assigned (eg. playwright `test` function that also has `test.describe`)
 * return the callable type without any other properties.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- this is the only way to define any function, unknown doesn't work
type OnlyFunction<T extends (...args: any) => unknown> = (
  ...args: Parameters<T>
) => ReturnType<T>

/**
 * The wrapper functions are defined in the caller file, to help playwright locate the tests in that file, not in node_modules/gherkin-wrapper/...
 *
 * @example
 * createBddWrapper(test, {
 *   describe: (...args) => test.describe(...args),
 *   test: (...args) => test(...args),
 * })
 */
export function createBddWrapper<T extends typeof baseTest>(
  test: T,
  {
    test: callTest,
    describe: callDescribe,
  }: { test: OnlyFunction<T>; describe: OnlyFunction<T["describe"]> },
) {
  // copy the caller function and set all the properties from the original
  const describeWrapper = Object.assign(callDescribe.bind({}), test.describe)

  const testWrapper = Object.assign(callTest.bind({}), test, {
    describe: describeWrapper,
  })

  // eslint-disable-next-line new-cap -- nothing we can do, external lib
  return new GherkinWrapper.forPlaywright(testWrapper)
}
```

**Comment by Niitch** [comment]
If I got it wright, what you want is to make playwright understand that the features/scenarios have been defined in your .spec.ts file, regardless that they have actually been built within the package.

In my opinion, the features/scenarios are defined in the feature file. I would prefer if playwright understood definitions from this  file instead. The only "code" for which playwright should care about its actual location is the actions defined in the .spec.ts file (using the given/when/then methods or hooks).

**I'm currently searching a way to alter the `location` objects playwright holds for each test suite/case/step.** This way I could force playwright to use the feature file lines for display in both the UI mode and the reporters.

<br>

The solution you suggest is a good workaround but for me has 2 main drawbacks:
- A call to `GherkinWrapper.forPlaywright` would less clear as you'll need a bind for more functions from the playwright `test` object. I want to keep this package usage as clear as possible. Espacially for this kind of feature that should work no matter what.
- You can't run a specific test from the feature.file (my solution neither), even if the playwright ui mode and the vscode test extension show options for it. A good adaptation would be to make them able to extract a partial `GherkinDocument` from the feature file and pass it to he wrapper

_NB: You can use `new GherkinWrapper.forPlaywright<T>(testWrapper)` in your example to help the ts intellisense._

<br>

About the vscode test extension, the test shows in it for me after running the test file from the test panel at least once.

**Comment by cassus** [comment]
> If I got it wright, what you want is to make playwright understand that the features/scenarios have been defined in your .spec.ts file, regardless that they have actually been built within the package.
> 
> In my opinion, the features/scenarios are defined in the feature file. I would prefer if playwright understood definitions from this file instead. The only "code" for which playwright should care about its actual location is the actions defined in the .spec.ts file (using the given/when/then methods or hooks).
>

Yes, that would be the ideal solution! I didn't know if it was even possible. :D  I like that @Niitch you are aiming high! I updated the **Expected behavior** section:

* Option A (ideal):  
   * If playwright understood the location of the test as the feature file and line where the Example / Scenario was specified
* Option B (has workaround):
   * If playwright understood the location of the test as the TS file where I called `wrapper.test`.



> **I'm currently searching a way to alter the `location` objects playwright holds for each test suite/case/step.** This way I could force playwright to use the feature file lines for display in both the UI mode and the reporters.
> 

That would be awesome!

---


> * You can't run a specific test from the feature.file (my solution neither), even if the playwright ui mode and the vscode test extension show options for it. A good adaptation would be to make them able to extract a partial `GherkinDocument` from the feature file and pass it to he wrapper

I'm not sure what you're referring to, I can run specific tests from a feature file in VSCode: https://recordit.co/9peyV0p1xZ (same in UI mode)

---

> _NB: You can use `new GherkinWrapper.forPlaywright<T>(testWrapper)` in your example to help the ts intellisense._

I'm on TS 5.0.4, strict mode, and all was inferred correctly without the need to explicitly pass the generic

**Comment by Niitch** [comment]
I found that the playwright `location` objects are built from stack traces by the `test.describe`/`test`/`test.step` functions by alteration of the `Error.captureStackTrace` method.

I don't know yet how to alter the stacktrace itself, if it is even possible.

As a workaround for the reporting, I made an extension of the playwright internal HTML reporter. You can build one as shown bellow and use it as a custom reporter.
```typescript
import GherkinWrapper from "../../src"

export default GherkinWrapper.forPlaywright.htmlReporter(path.resolve("path/to/file.feature"))
```
⚠️ __*You will need to expose manually the playwright built-in HTML reporter from its package*__ ⚠️
For now, it works only for a single feature file. If no better solution is found in the future, I'll implement a better version of it and extend the logic to every playwright built-in reporters.

I'll open a issue on the playwright repo about all this.

**Comment by NikkTod** [comment]
Hey guys could you please help me. I have also similar problem, when I run `npx playwright test example.spec.ts` it says no test found.

I have the following structure:
![image](https://github.com/Niitch/gherkin-wrapper/assets/94455079/e12610ca-580a-4b77-ae77-7eec2b641d33)


example.spec.ts

```
import { test as base } from "@playwright/test";
import GherkinWrapper from "gherkin-wrapper";

// Add fixtures
const test = base.extend<{value: string}>({
    value: async ({}, use) => {
        await use('go')
    }
})

// Build the wrapper
const wrapper = new GherkinWrapper.forPlaywright(test)


// Register step functions
wrapper.given(/I open url "(.*)"/, async ({ page, value }, { match }) => {
    console.log(match)
    await page.goto(value);
})

// Run tests 
wrapper.test('./feature/example.feature')
```

playwright.config

```
export default defineConfig({
  testDir: './step-definitions',
  ......
```

**Comment by cassus** [comment]
Hey @NikkTod , I had the same issue and found two workarounds:

- the code above https://github.com/Niitch/gherkin-wrapper/issues/4#issuecomment-1546982920
- use `--grep` on the name of the describe / test instead of filtering on the file name

**Comment by NikkTod** [comment]
Thanks @cassus , I tried running it with grep and it worked 💯

**Comment by NikkTod** [comment]
So my understanding is that for now we have workaround for running tests from terminal. (e.g npx playwright test --grep "Check title")

What is the situation with the pw vscode extention and UI mode? 

![image](https://github.com/Niitch/gherkin-wrapper/assets/94455079/29201e1c-0eec-46fc-a9cb-e1d1a89d8bb1)

**Comment by Niitch** [comment]
I've found a way to make it works for both the UI mode and the reporter.

With this solution you will have to choose between using the ideal behaviour and using the vscode test extension. They don't work together.

**Comment by NikkTod** [comment]
@Niitch by "ideal behaviour" you mean we have to choose between UI mode and vscode test extention?

**Comment by Niitch** [comment]
"ideal behaviour" refers to "Playwright understood the location of the test as the feature file and line where the Example / Scenario was specified"

For now, if you want to use the vscode test extension, you can do it using the workaround at best. The extension doesn't launch the tests at all when I'm using the solution I've found.

**Comment by NikkTod** [comment]
Hi @Niitch, sorry to bother, is there a chance we have the Vs Code extention, UI mode and html report working?

**Comment by Niitch** [comment]
@NikkTod
I merged and published  the feature for the ui mode and the reports just now (v0.3.0).

About the VSCode test explorer, I have found it would be better to have a dedicated extension for the gherkin-wrapper.
The [cucumber-quick](https://github.com/abhinaba-ghosh/cucumber-quick) project did it great for cucumber.
